### ISSUE TITLE
Duplicated QoS2 messages 

### ISSUE THREAD

**Issue by jsaak** [eclipse/mosquitto#1751]
We use a mosquitto broker, and a connected client using the mosquitto lib. (using the 1.5.x branch)
And very rarely we see QoS2 messages appearing twice.
Unfortunately i can not reproduce it, so I started to delve into the code.

I do not understand, how it is working:
Is it true the in case of network lag - when broker sends it twice, once normally, and then with DUP set to true - the PUBLISH message will be stored multiple times in mosq_md_in? (in lib/handle_publish.c:116)

And when PUBREL it gets removed only once?  (in lib/handle_pubrel.c:69)

**Comment by ralight** [comment]
On a first glance it appears as though you may be correct that duplicate message IDs would cause a problem, I'll have to try and test it properly.

In terms of the failure cause, neither broker not client should send a duplicate PUBLISH just due to lag. The only time when resending is allowed is on reconnection.

Are your duplicates close by to one another in time, or distant?

**Comment by jsaak** [comment]
It was close.
The client was overwhelmed and could not process the first publish.

```
1593586652: Sending PUBLISH to BS3_live (d0, q2, r0, m19515, 'BA/SL/BA0398/bikepos', ... (48 bytes))
1593586673: Sending PUBLISH to BS3_live (d1, q2, r0, m19515, 'BA/SL/BA0398/bikepos', ... (48 bytes))
```

**Comment by ralight** [comment]
21 seconds between messages suggests it is the old retry_interval behaviour from 1.4.x and earlier that is causing this on the broker side - it was set to 20 seconds. Are you sure the broker is up to date? Getting rid of the duplicate message would solve the problem for you in the short term while the library is fixed.

**Comment by jsaak** [comment]
You are right, we use 1.4.10 as the broker (I thought we upgraded all of them, but it seems not)
So you removed retry_interval all-together?

Still if using a different broker, they may be using this functionality. So a client fix is needed. The cleaner solution would be to check for the message_id before storing. I am not sure how slow it would be.
Or maybe a warning should be enough, to make sure that the broker does not send duplicate packages at all?

**Comment by ralight** [comment]
The client fix is absolutely needed.

Sending retries is now explicitly only allowed on reconnect by the MQTT spec, so I would be surprised if any broker implemented retries on timeout.
